### PR TITLE
[Fix] Per-rank SHM + streaming 1× host peak + FX device placement

### PR DESCRIPTION
**Pull request by cennn**
## Motivation

`model_cpu_offload` materializes weights into `/dev/shm` so ranks on one node can share pinned pages. On `main`, rank 0 writes a single flat file per dtype and every rank maps it — correct when all ranks hold the same weights, but when ranks hold different shards (expert parallelism), every shard is silently replaced with rank 0's data. MoE routing still runs, experts are numerically wrong, producing garbled video.

The same `main` path allocates a full `flat_buffer = torch.zeros(total_numel)` before writing to disk, so host anonymous memory peaks at ~2× model size. For a 200 GB+ MoE model across 8 ranks this means ~700 GB peak — dangerously close to host OOM.

Separately, `main` only rewrites `device=` kwargs on factory functions (`torch.empty`, `torch.zeros`, …) in the FX graph. Offload tracing also emits `.to(device("cpu"))` call_method nodes and leaves CPU `example_value` metadata on arbitrary nodes, which Inductor cannot handle — it raises `FakeTensor device mismatch` on ops like `aten.index_select` receiving both cuda and cpu arguments.

### 1. Auto-detect per-rank SHM via weight fingerprint

Each rank computes a SHA256 fingerprint (param names + shapes + sampled head/tail data, < 1 s for any model size), then `dist.all_gather` compares across ranks. All identical → single shared mmap; any differ → per-rank mmap. Zero configuration needed for both DP/CP and EP scenarios.

Override: `MAGI_COMPILE_OFFLOAD_CONFIG__FORCE_PER_RANK_WEIGHTS` env var (`1`/`0`) bypasses auto-detection when needed.

### 2. Streaming copy-and-replace (1× host peak)

`_stream_copy_and_replace` copies each parameter into the mmap and immediately replaces it in the module, so only one parameter's worth of duplication exists at any moment. The old `flat_buffer` + `tofile()` + `load_state_dict` path is eliminated.

| Path | Anonymous growth | vs model |
| --- | ---: | ---: |
| `main` (`flat_buffer`) | 260 MB | **1.02×** |
| this PR (streaming) | 4 MB | **0.02×** |

*(measured via `smaps_rollup Anonymous` on 64 MB model in isolated subprocess)*

### 3. Full FX device placement rewrite

`fix_graph_device_placement` (module-level pure function) rewrites:
- `call_function` nodes with `device=cpu` kwargs (covers all ops, not just factory functions)
- `call_method` `.to(cpu)` nodes (positional args and kwargs)
- All CPU `example_value` metadata, recursively through nested list/tuple

## Other

- `offload()` recursively penetrates plain objects' `__dict__` to move CUDA tensors to CPU, skipping `nn.Module` to avoid interfering with model internals.
- `_force_cpu` skips the GPU round-trip when the tensor is already on CPU and `fn` does not change dtype.
- `host_memory.py`: lightweight `/proc`-based host memory introspection (`fmt_host_mem()`), logged at key offload stages.
- `OffloadConfig.force_per_rank_weights`: `None` (auto-detect) / `True` / `False`, exposed as Pydantic field with env var support via `BaseSettings`.

## Evidence

- 8×5090 base compile+offload warmup completes on the SHM path.
- All-5090 E2E (textenc CPU + base/sr/vae 5090) produced valid videos from `samples_5req_diverse.json`.

> [attachment: garbled vs fixed video comparison, base warmup logs — to be pasted]

## Tests (+701 lines, 3 new files)

- `test_ep_shared_memory.py`: 2-rank gloo repro — `per_rank=False` corrupts shards, `per_rank=True` preserves; 4 fingerprint auto-detection tests.
- `test_fix_to_cpu_in_graph.py`: FX graph `.to(cpu)` + `device=cpu` rewrite — metadata-only fix fails, full rewrite succeeds.
- `test_shm_memory_peak.py`: subprocess-isolated `smaps_rollup Anonymous` measurement — batch shows ~1× growth, streaming shows ~0× growth.

### CODE REVIEW

File: magi_compiler/_api.py [hunk 1]
```diff
@@ -532,12 +648,29 @@ def _cpu_apply(self, fn):
...
+                if target_dt == dt:
+                    return t
+                return t.to(dtype=target_dt)
             return fn(t).cpu()
 
         _orig_apply(self, _force_cpu)
+        magi_logger.info('[offload] after _force_cpu: %s', fmt_host_mem())
 
         # create shared memory tensors for all parameters/buffers on CPU
+        ep_size = get_topology_dim("ep")
```
Comment (jiahy0825) [inline magi_compiler/_api.py]: The compiler should not be aware of framework-level concepts. That feels too hacky.

**Review by jiahy0825 — APPROVED** [review]
LGTM
